Index: build/venv/lib/python3.8/site-packages/magenta/models/onsets_frames_transcription/model_tpu.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/build/venv/lib/python3.8/site-packages/magenta/models/onsets_frames_transcription/model_tpu.py b/build/venv/lib/python3.8/site-packages/magenta/models/onsets_frames_transcription/model_tpu.py
--- a/build/venv/lib/python3.8/site-packages/magenta/models/onsets_frames_transcription/model_tpu.py	
+++ b/build/venv/lib/python3.8/site-packages/magenta/models/onsets_frames_transcription/model_tpu.py	(date 1657105944168)
@@ -312,7 +312,7 @@
   return frame_logits, onset_logits, offset_logits, velocity_values
 
 
-def model_fn(features, labels, mode, params, config):
+def model_fn(features, labels, mode, params, config, include_metrics=True):
   """Builds the acoustic (for Estimator API)."""
   del config
   hparams = params
@@ -334,7 +334,7 @@
 
   return estimator_spec_util.get_estimator_spec(
       hparams, mode, features, labels, frame_logits, onset_logits,
-      offset_logits, velocity_values, offset_network=hparams.offset_network)
+      offset_logits, velocity_values, offset_network=hparams.offset_network, include_metrics=include_metrics)
 
 
 def get_default_hparams():
